R/module_bookmark_manager.R

```diff
@@ -6,196 +6,137 @@
-#' either by setting an argument in a `shinyApp` call or by calling a special function. In `teal` bookmarks are enabled
-#' by automatically setting an option when the package is loaded.
+#' `shinyOptions("bookmarkStore" = "server")` is set in `teal` by default on package load. Using the
+#' `url` option is not supported.
 #'
-#' If the option is set to a different value by the app developer, the bookmark manager will forbid creating bookmarks.
+#' `teal` does not guarantee that all `teal_module` elements are bookmarkable. Those that are, have the
+#' `teal_bookmarkable` attribute set to `TRUE`. The bookmark manager modal displays a warning if any
+#' modules are not bookmarkable. In order to set external `teal_module` elements as bookmarkable, the
+#' `teal_bookmarkable` attribute must be set to `TRUE` by the module developer.
 #'
-#' @param id (`character(1)`) `shiny` module instance id.
+#' @inheritParams module_wunder_bar
 #'
 #' @return `reactiveVal` containing a named list of bookmark URLs.
 #'
-#' @name module_bookmark_manager
 #' @aliases bookmark bookmark_manager bookmark_manager_module
 #'
-
 #' @rdname module_bookmark_manager
 #' @keywords internal
 #'
 bookmark_manager_ui <- function(id) {
   ns <- NS(id)
-  div(
-    class = "manager_content",
-    div(
-      class = "manager_table_row",
-      span(tags$b("Bookmark manager")),
-      actionLink(ns("bookmark_add"), NULL, icon = suppressMessages(icon("solid fa-bookmark")), title = "add bookmark"),
-      NULL
-    ),
-    uiOutput(ns("bookmark_list"))
+  tags$button(
+    id = ns("do_bookmark"),
+    class = "btn action-button wunder_bar_button",
+    title = "Add bookmark",
+    tags$span(
+      suppressMessages(icon("solid fa-bookmark")),
+      uiOutput(ns("bookmark_warning"), inline = TRUE)
+    )
   )
 }
 
 #' @rdname module_bookmark_manager
 #' @keywords internal
 #'
-bookmark_manager_srv <- function(id) {
+bookmark_manager_srv <- function(id, modules) {
   checkmate::assert_character(id)
-
+  checkmate::assert_class(modules, "teal_modules")
   moduleServer(id, function(input, output, session) {
     logger::log_trace("bookmark_manager_srv initializing")
-
-    # Enforce server-side bookmarking ----
+    ns <- session$ns
     bookmark_option <- getShinyOption("bookmarkStore", "disabled")
-    if (bookmark_option != "server") {
-      shinyjs::disable("bookmark_add")
-      output$bookmark_list <- renderUI({
-        div(
-          class = "manager_placeholder",
-          sprintf("Bookmarking has been set to \"%s\".", bookmark_option), tags$br(),
-          "Only server-side bookmarking is supported.", tags$br(),
-          "Please contact your app developer."
+    is_unbookmarkable <- rapply(
+      modules_bookmarkable(modules),
+      Negate(isTRUE),
+      how = "unlist"
+    )
+    # Render bookmark warnings count
+    output$bookmark_warning <- renderUI({
+      if (!identical(bookmark_option, "server")) {
+        shinyjs::hide("do_bookmark")
```

Comment:
Hiding the button renders the warning in `modal_content` obsolete, doesn't it?

Reply:
This thing is not in a modal - this is a warning badge. I made small changes anyway

Reply:
My mistake.